File: 03-Arrays-Part-III/python/06_reverse_pairs.py
class Solution:
    def reversePairs(self, nums: List[int]) -> int:
        return self.mergeSort(nums, 0, len(nums) - 1)

    def mergeSort(self, nums, left, right):
        if left >= right:
            return 0

        mid = left + (right - left) // 2
        count = self.mergeSort(nums, left, mid) + self.mergeSort(nums, mid + 1, right)

        j = mid + 1
        for i in range(left, mid + 1):
            while j <= right and nums[i] > 2 * nums[j]:
                j += 1
            count += (j - (mid + 1))

        temp = []
        p1, p2 = left, mid + 1
        while p1 <= mid and p2 <= right:
            if nums[p1] <= nums[p2]:
                temp.append(nums[p1])
                p1 += 1
            else:
                temp.append(nums[p2])
                p2 += 1
        while p1 <= mid:
            temp.append(nums[p1])
            p1 += 1
        while p2 <= right:
            temp.append(nums[p2])
            p2 += 1

        for i in range(len(temp)):
            nums[left + i] = temp[i]

        return count

File: 03-Arrays-Part-III/cpp/06-reverse-pairs.cpp
class Solution {
public:
    int reversePairs(vector<int>& nums) {
        return mergeSort(nums, 0, nums.size() - 1);
    }

private:
    int mergeSort(vector<int>& nums, int left, int right) {
        if (left >= right) return 0;

        int mid = left + (right - left) / 2;
        int count = mergeSort(nums, left, mid) + mergeSort(nums, mid + 1, right);

        int j = mid + 1;
        for (int i = left; i <= mid; i++) {
            while (j <= right && nums[i] > 2LL * nums[j]) {
                j++;
            }
            count += (j - (mid + 1));
        }

        vector<int> temp(right - left + 1);
        int p1 = left, p2 = mid + 1, k = 0;
        while (p1 <= mid && p2 <= right) {
            if (nums[p1] <= nums[p2]) {
                temp[k++] = nums[p1++];
            } else {
                temp[k++] = nums[p2++];
            }
        }
        while (p1 <= mid) temp[k++] = nums[p1++];
        while (p2 <= right) temp[k++] = nums[p2++];

        for (int i = 0; i < (int)temp.size(); i++) {
            nums[left + i] = temp[i];
        }
        return count;
    }
};
